cilium-agent crash and unable to recover when `enable-l7-proxy` is toggled

### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Cilium-agent crashes and unable to recover, as explained below,

- When "enable-l7-proxy: true" and using DNS proxy to add L3/L4 rules for the toFQDNs, the FQDN cache gets populated as lookup entries and the policies were working ok. 
- When "enable-l7-proxy: false" on the same daemonset and rolling restart the cilium agents, cilium-agent was healthy for a brief moment until the "FQDN garbage collector" kicked in, after which it bails out on "panic: runtime error: invalid memory address or nil pointer dereference". 
- Noticed that this happened only on some of the nodes, the ones that had entries in the FQDN cache, and those nodes never recovered. 
- Also noticed that the same behaviour can be reproduced by `cilium fqdn cache clean` which forces the crash.
- FQDN garbage collector was started on the nodes that had been crashing, whereas on the other nodes only the conntrack garbage collector was running from startup. 
- Reverting to "enable-l7-proxy: true", all the nodes were back healthy.

cc: @supreethrao


### Cilium Version

Versions:
- 1.9.8
- 1.10.5

### Kernel Version

Linux 5.4.0-1054-aws #57~18.04.1-Ubuntu SMP x86_64 GNU/Linux

### Kubernetes Version

Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.0", GitCommit:"9e991415386e4cf155a24b1da15becaa390438d8", GitTreeState:"clean", BuildDate:"2020-03-25T14:58:59Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"darwin/amd64"}

Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.14", GitCommit:"d2a081c8e14e21e28fe5bdfa38a817ef9c0bb8e3", GitTreeState:"clean", BuildDate:"2020-08-13T12:24:51Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}


### Sysdump

_No response_

### Relevant log output

```shell
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x221b242]

goroutine 1953 [running]:
github.com/cilium/cilium/pkg/proxy.(*Proxy).UseCurrentNetworkPolicy(0x0, 0x2f450f0, 0xc000efc000, 0xc000d35bc0, 0xc0027e1300)
        /go/src/github.com/cilium/cilium/pkg/proxy/proxy.go:656 +0x22
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).useCurrentNetworkPolicy(0xc000efc000, 0xc0027e1300)
        /go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:167 +0x8f
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).ApplyPolicyMapChanges(0xc000efc000, 0xc0027e1300, 0x0, 0x0)
        /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:1217 +0x138
github.com/cilium/cilium/pkg/endpointmanager.(*EndpointManager).UpdatePolicyMaps.func2(0xc000702570, 0xc0027e1300, 0xc000702580, 0xc000efc000)
        /go/src/github.com/cilium/cilium/pkg/endpointmanager/manager.go:174 +0x65
created by github.com/cilium/cilium/pkg/endpointmanager.(*EndpointManager).UpdatePolicyMaps
        /go/src/github.com/cilium/cilium/pkg/endpointmanager/manager.go:171 +0x1ae
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x221b242]

goroutine 1954 [running]:
github.com/cilium/cilium/pkg/proxy.(*Proxy).UseCurrentNetworkPolicy(0x0, 0x2f450f0, 0xc000efc900, 0xc001798140, 0xc0027e1300)
        /go/src/github.com/cilium/cilium/pkg/proxy/proxy.go:656 +0x22
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).useCurrentNetworkPolicy(0xc000efc900, 0xc0027e1300)
        /go/src/github.com/cilium/cilium/pkg/endpoint/policy.go:167 +0x8f
github.com/cilium/cilium/pkg/endpoint.(*Endpoint).ApplyPolicyMapChanges(0xc000efc900, 0xc0027e1300, 0x0, 0x0)
        /go/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:1217 +0x138
github.com/cilium/cilium/pkg/endpointmanager.(*EndpointManager).UpdatePolicyMaps.func2(0xc000702570, 0xc0027e1300, 0xc000702580, 0xc000efc900)
        /go/src/github.com/cilium/cilium/pkg/endpointmanager/manager.go:174 +0x65
created by github.com/cilium/cilium/pkg/endpointmanager.(*EndpointManager).UpdatePolicyMaps
        /go/src/github.com/cilium/cilium/pkg/endpointmanager/manager.go:171 +0x1ae
```


### Anything else?

Based on the source, it looks like,
- envoy control plane XDS server is started only if L7 proxy is enabled. https://github.com/cilium/cilium/blob/fcd00390c30c6eeffbe2fefa81d5b22e59397297/daemon/cmd/daemon.go#L579
- this condition should be calling `!isProxyDisabled()` on the endpoint instead of doing a nil check on the interface because the type exists but value is nil, 
https://github.com/cilium/cilium/blob/c78195592e65349c72724784c834cd8cd5b20218/pkg/endpoint/policy.go#L154 


### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

--- a/pkg/endpoint/policy.go
+++ b/pkg/endpoint/policy.go
@@ -151,7 +151,7 @@ func (e *Endpoint) useCurrentNetworkPolicy(proxyWaitGroup *completion.WaitGroup)
 		return
 	}
 
-	if e.proxy != nil {
+	if !e.isProxyDisabled() {
 		// Wait for the current network policy to be acked
 		e.proxy.UseCurrentNetworkPolicy(e, e.desiredPolicy.L4Policy, proxyWaitGroup)
 	}

